### PR TITLE
Implement front end module permissions

### PR DESCRIPTION
**Pull request by bezin**
This PR implements #6016: Same permission management for front-end modules as we have for content elements. 

See this short video https://github.com/contao/contao/assets/4400435/b5722372-b6b7-4537-9ed5-d6ddfccfa469

Not yet finished, but I think now is the time to maybe discuss details :-)

### CODE REVIEW

**Comment by aschempp** [comment]
Awesome! 🎉 
Regarding the module access permissions, you should use a security voter instead of a listener on the DCA onload callback.

**Comment by aschempp** [comment]
Looks very good to me! Haven't manually tested it yet though 🙃

File: core-bundle/tests/Security/Voter/BackendAccessVoterTest.php [hunk 1]
```diff
@@ -167,6 +167,12 @@ public function userDataProvider(): \Generator
             'text',
         ];
 
+        yield 'Check access on front end module' => [
+            ['frontendModules' => ['navigation']],
+            ContaoCorePermissions::USER_CAN_ACCESS_FRONTEND_MODULE_TYPE,
```
Comment (leofeyer) [inline core-bundle/tests/Security/Voter/BackendAccessVoterTest.php:172]: We should replace `FRONTEND_MODULE` with `MODULE` and also `$frontendModule`, `frontendModules` etc. everywhere else, because we use `module` and `element` internally. The same constant for content elements is also called `USER_CAN_ACCESS_ELEMENT_TYPE` and not `USER_CAN_ACCESS_CONTENT_ELEMENT_TYPE`.

File: tools/install/composer.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (leofeyer) [inline tools/install/composer.json]: What is `tools/install`?

File: core-bundle/contao/dca/tl_module.php [hunk 3]
```diff
@@ -146,7 +146,7 @@
 			'options_callback'        => array('tl_module', 'getModules'),
 			'reference'               => &$GLOBALS['TL_LANG']['FMD'],
 			'eval'                    => array('helpwizard'=>true, 'chosen'=>true, 'submitOnChange'=>true, 'tl_class'=>'w50'),
-			'sql'                     => "varchar(64) COLLATE ascii_bin NOT NULL default 'navigation'"
+			'sql'                     => array('name'=>'type', 'type'=>'string', 'length'=>64, 'default'=>'navigation', 'customSchemaOptions'=>array('collation'=>'ascii_bin'))
```
Comment (leofeyer) [inline core-bundle/contao/dca/tl_module.php]: Why this change?

File: core-bundle/src/Security/Voter/DataContainer/FrontendModulesVoter.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+            $isGranted = match (true) {
+                $subject instanceof ReadAction => true,
+                $subject instanceof CreateAction,
+                $subject instanceof UpdateAction,
+                $subject instanceof DeleteAction => $this->isAllowedType($subject, $user),
+                default => false,
+            };
+
+            return $isGranted ? self::ACCESS_GRANTED : self::ACCESS_DENIED;
```
Comment (leofeyer) [inline core-bundle/src/Security/Voter/DataContainer/FrontendModulesVoter.php]: This is wrong, because our voters must never grant access. You should use the new [`AbstractDataContainerVoter`](https://github.com/contao/contao/blob/5.x/core-bundle/src/Security/Voter/DataContainer/AbstractDataContainerVoter.php), which will fix this issue.

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/tests/Security/Voter/BackendAccessVoterTest.php (hunk 1 above)
Comment (bezin) [reply core-bundle/tests/Security/Voter/BackendAccessVoterTest.php:172]: The issue is, that `module` is internally used for back-end modules, too. The field `tl_user.modules` is already taken. This would need to be `frontendModules`. Otherwise I am fine with that.

File: tools/install/composer.json (hunk 2 above)
Comment (bezin) [reply tools/install/composer.json]: That was by accident. Removed in eeae688f32

File: core-bundle/tests/Security/Voter/BackendAccessVoterTest.php (hunk 1 above)
Comment (bezin) [reply core-bundle/tests/Security/Voter/BackendAccessVoterTest.php:172]: Same goes for the Constants in `ContaoCorePermissions`: There is already `USER_CAN_ACCESS_MODULE` for back-end modules

File: core-bundle/contao/dca/tl_module.php (hunk 3 above)
Comment (bezin) [reply core-bundle/contao/dca/tl_module.php]: Fixed in [0b8bb4f](https://github.com/contao/contao/pull/6232/commits/0b8bb4f887371998b6ab5d5efbe2a52d2b74d0b6)

File: core-bundle/src/Security/Voter/DataContainer/FrontendModulesVoter.php (hunk 4 above)
Comment (bezin) [reply core-bundle/src/Security/Voter/DataContainer/FrontendModulesVoter.php]: Implemented the new abstract voter in dcc3e17b9d3349d42e961d115d1651c40b201c55

File: core-bundle/tests/Security/Voter/BackendAccessVoterTest.php (hunk 1 above)
Comment (leofeyer) [reply core-bundle/tests/Security/Voter/BackendAccessVoterTest.php:172]: Hm, that is unfortunate. @contao/developers What do you think? Should we use `USER_CAN_ACCESS_FRONTEND_MODULE_TYPE` then or maybe `USER_CAN_ACCESS_FE_MODULE_TYPE`?

Comment (Toflar) [reply core-bundle/tests/Security/Voter/BackendAccessVoterTest.php:172]: I would use `FRONTEND_MODULE`, not `FE`. We don't use abbreviations other than `DBAFS` imho. We already have `USER_CAN_ACCESS_FRONTEND_MODULES` for the theme permissions and we did not name it `FE` there either.

Comment (leofeyer) [reply core-bundle/tests/Security/Voter/BackendAccessVoterTest.php:172]: That is true. 👍

**Review by leofeyer — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bezin.
